lib/targets/cuda/quda_api.cpp

```diff
@@ -54,9 +123,10 @@ namespace quda
   static TimeProfile apiTimer("CUDA API calls (runtime)");
 #endif
 
-  qudaError_t qudaLaunchKernel(const void *func, const TuneParam &tp, void **args, qudaStream_t stream)
+  qudaError_t qudaLaunchKernel(const void *func, const TuneParam &tp, const qudaStream_t &stream, const void *arg)
   {
-    if (tp.set_max_shared_bytes) {
+    // if launch requests the maximum shared memory and the device supports it then opt in
+    if (tp.set_max_shared_bytes && device::max_dynamic_shared_memory() > device::max_default_shared_memory()) {
```

Comment:
Design questions, should we replace the `qudaFunc*` routines below with `cudaFunc*`?